hands-on-exercise/4-run-in-prod/2-migration-info.md

```diff
@@ -0,0 +1,2117 @@
+2. Calling `.GetPlayerInfo` and `.SetPlayerInfo` twice per game just to do `+1` adds up quickly. Remember that both of these calls are database calls. You could be facing a 12-hour job, during which your chain is offline.
+3. Doing it all in a sequential manner would take even more time, as each blocking call blocks the whole process.
+
+### Proposed solution
+
+Fortunately, there exist ways to mitigate these limitations:
+
+1. You do not need to get all the games at once. The [`keeper.StoredGameAll`](https://github.com/cosmos/b9-checkers-academy-draft/blob/player-info-handling/x/checkers/keeper/grpc_query_stored_game.go#L14) function offers pagination. With this, you can limit the impact on the RAM requirement, at the expense of multiple queries.
+2. Within each subset of games, you can compute in memory the player list and how many wins and losses each player has. With this _mapping_ done, you can add the (in-memory) intermediary `WonCount` and `LostCount` sums to each player's stored sums. With this, a `+1` is potentially replaced by a `+k`, at once reducing the number of calls to `.GetPlayerInfo` and `.SetPlayerInfo`.
+3. You can separate the different calls and computations into [Go routines](https://gobyexample.com/goroutines) so that a blocking call does not prevent other computations from taking place in the meantime.
+
+The routines will use **channels** to communicate between themselves and the main function:
+
+1. A _stored-game_ channel, that will pass along chunks of games in the `[]types.StoredGame` format.
+2. A _player-info_ channel, that will pass along intermediate computations of player information in the simple `types.PlayerInfo` format.
+3. A _done_ channel, whose only purpose is to flag to the main function when all has been processed.
+
+Each channel should also be able to pass an optional error, so tuples will be used.
+
+The processing **routines** will be divided as per the following:
+
+1. The **game loading** routine will:
+
+    * Fetch all games in paginated arrays.
+    * Send the separate arrays on the _stored-game_ channel.
+    * Send an error on the _stored-game_ channel if any is encountered.
+    * Close the _stored-game_ channel after the last array or on an error.
+
+2. The **game processing** routine will:
+    
+    * Receive separate arrays of games from the _stored-game_ channel.
+    * Compute the aggregate player info records from them. I.e. **_map_**.
+    * Send the results on the _player-info_ channel.
+    * Pass along an error if it receives any.
+    * Close the _player-info_ channel after the last stored game, or on an error.
+
+3. The **player info processing** routine will:
+
+    * Receive individual player info records from the _player-info_ channel.
+    * Fetch the existing (or not) corresponding player info from the store.
```

Comment:
@xavierlepretre I have ambiguity problems here. Possible meanings: 

1. Fetch the existing _(or **do not fetch** the existing)_ corresponding player info...
2. Fetch the existing _(or **the non-existent**)_ corresponding player info...

Both seem to be contradictory. The first because (I assume) a single instruction can't direct to do two opposite actions; the second because you can't fetch something that doesn't exist (unless "fetching" can also cause a new player info dataset to be created).

IF what we're saying is _"Fetch the info if it exists, and do not fetch anything if it doesn't"_ then this is an easy fix for me.

Reply:
It's version 2. I'd go with:

> Fetch the corresponding player info from the store. If it does not exist yet, it will create an empty new one.